### ISSUE TITLE
App crash due to MoPub SDK

### ISSUE THREAD

**Issue by theaterears-dev** [mopub/mopub-ios-sdk#267]
One line summary of the issue here

- [YES] I am submitting a bug report for existing functionality
- [YES ] I visited https://developers.mopub.com/ and found no answer
- [YES] I checked https://twittercommunity.com/c/advertiser-api/mopub to make sure that this issue has not already been filed
- [YES] I checked to make sure that this issue has not already been filed

#### Steps to reproduce the behavior:
App crashes when application is launched.

#### Expected behavior:
Application should not crash

#### Observed behavior:
Application crashes on launch.

Thousands of users are facing this issues and all point towards at [MPURLRequest initWithURL:]

#### Evidence:
#0. Crashed: com.apple.main-thread
0  libobjc.A.dylib                0x1ebe67be8 objc_retain + 8
1  TheatreEars                    0x102425824 -[MPURLRequest initWithURL:] (MPURLRequest.m:59)
2  TheatreEars                    0x10241ea14 +[MPSessionTracker trackEvent] (MPSessionTracker.m:38)
3  CoreFoundation                 0x1ecbe55bc __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 20
4  CoreFoundation                 0x1ecbe5588 ___CFXRegistrationPost_block_invoke + 64
5  CoreFoundation                 0x1ecbe4a7c _CFXRegistrationPost + 392
6  CoreFoundation                 0x1ecbe4728 ___CFXNotificationPost_block_invoke + 96
7  CoreFoundation                 0x1ecb5e524 -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1496
8  CoreFoundation                 0x1ecbe41d8 _CFXNotificationPost + 696
9  Foundation                     0x1ed5cc814 -[NSNotificationCenter postNotificationName:object:userInfo:] + 68
10 UIKitCore                      0x2191cec58 -[UIApplication _sendWillEnterForegroundCallbacks] + 232
11 UIKitCore                      0x218a7aa44 -[__UICanvasLifecycleMonitor_Compatability activateEventsOnly:withContext:completion:] + 2036
12 UIKitCore                      0x218a789c4 __82-[_UIApplicationCanvas _transitionLifecycleStateWithTransitionContext:completion:]_block_invoke + 744
13 UIKitCore                      0x218a7868c -[_UIApplicationCanvas _transitionLifecycleStateWithTransitionContext:completion:] + 428
14 UIKitCore                      0x218a7d1cc __125-[_UICanvasLifecycleSettingsDiffAction performActionsForCanvas:withUpdatedScene:settingsDiff:fromSettings:transitionContext:]_block_invoke + 220
15 UIKitCore                      0x218a7dfb0 _performActionsWithDelayForTransitionContext + 112
16 UIKitCore                      0x218a7d084 -[_UICanvasLifecycleSettingsDiffAction performActionsForCanvas:withUpdatedScene:settingsDiff:fromSettings:transitionContext:] + 244
17 UIKitCore                      0x218a81d84 -[_UICanvas scene:didUpdateWithDiff:transitionContext:completion:] + 360
18 UIKitCore                      0x218db13b8 -[UIApplicationSceneClientAgent scene:handleEvent:withCompletion:] + 464
19 FrontBoardServices             0x1ef5ef920 __80-[FBSSceneImpl updater:didUpdateSettings:withDiff:transitionContext:completion:]_block_invoke_3 + 224
20 libdispatch.dylib              0x1ec6b57d4 _dispatch_client_callout + 16
21 libdispatch.dylib              0x1ec65a5d8 _dispatch_block_invoke_direct$VARIANT$mp + 224
22 FrontBoardServices             0x1ef6290bc __FBSSERIALQUEUE_IS_CALLING_OUT_TO_A_BLOCK__ + 40
23 FrontBoardServices             0x1ef628d58 -[FBSSerialQueue _performNext] + 408
24 FrontBoardServices             0x1ef629310 -[FBSSerialQueue _performNextFromRunLoopSource] + 52
25 CoreFoundation                 0x1ecc062bc __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 24
26 CoreFoundation                 0x1ecc0623c __CFRunLoopDoSource0 + 88
27 CoreFoundation                 0x1ecc05b74 __CFRunLoopDoSources0 + 256
28 CoreFoundation                 0x1ecc00a60 __CFRunLoopRun + 1004
29 CoreFoundation                 0x1ecc00354 CFRunLoopRunSpecific + 436
30 GraphicsServices               0x1eee0079c GSEventRunModal + 104
31 UIKitCore                      0x2191b7b68 UIApplicationMain + 212
32 TheatreEars                    0x10247e830 main (main.m:14)
33 libdyld.dylib                  0x1ec6c68e0 start + 4

**Comment by chauduyphanvu1** [comment]
@theaterears-dev, from the description, it sounds like the app is crashing when initializing the MoPub SDK (since you mention app launch). If you could tell us a bit more about what the app is doing (or confirm that hypothesis), we should be able to debug the crash. For now, try to comment out the [initialization code](https://github.com/mopub/mopub-ios-sdk/blob/master/MoPubSampleApp/AppDelegate.m#L43-L48) and see if the crash persists. From code it seems like it's crashing when trying to do work with the [user agent](https://github.com/mopub/mopub-ios-sdk/blob/5.5.0/MoPubSDK/Internal/MPURLRequest.m#L59). 

Vu Chau
MoPub SDK Team

**Comment by chauduyphanvu1** [comment]
> Do you think by making the respective ad delegates nil in my dealloc method might solve the problem?

I don't think that's a solution because your stack trace indicates a code path that triggers session tracking (and crashes in the attempt). Session tracking is performed when the MoPub SDK is initialized for that app session, and that has to happen before ad requests (and ad shows). We can safely say there's no connection between your ad delegates and this crash.

In your upcoming app release, would it be convenient if you add a log to the SDK (the code is modifiable if you integrate via CocoaPods) to print the user agent (or to forward that information to you somehow)? [This line](https://github.com/mopub/mopub-ios-sdk/blob/5.5.0/MoPubSDK/Internal/MPURLRequest.m#L59) does that. I'm curious if the user agent is broken when it crashes. Thank you!

**Comment by theaterears-dev** [comment]
I haven't used CocaPods to integrate MoPub, instead manually added MoPub folder to my project by downloading from git.
Do you think this could also be one of the reason?

When I am using my test devices, the debugger is able to log MoPub UserAgent in the console. I'll try and figure out a way to log this on the live build.

Also I noticed in my Crashlytics, that there are multiple different session where the crash from somewhere else such as requesting Native Ads or Interstitial Ads starts but ends up at [MPURLRequest initWithUrl].

**More Evidence**:

Crashed: com.apple.main-thread
0  libobjc.A.dylib                0x1ac55bbe8 objc_retain + 8
1  TheatreEars                    0x1047b9824 -[MPURLRequest initWithURL:] (MPURLRequest.m:59)
2  TheatreEars                    0x1048b17b0 -[MPAdServerCommunicator loadURL:] (MPAdServerCommunicator.m:90)
3  TheatreEars                    0x1046fb1a4 -[MPNativeAdRequest loadAdWithURL:] (MPNativeAdRequest.m:146)
4  TheatreEars                    0x1046fae18 -[MPNativeAdRequest startForAdSequence:withCompletionHandler:] (MPNativeAdRequest.m:114)
5  TheatreEars                    0x1048781fc -[MPNativeAdSourceQueue replenishCache] (MPNativeAdSourceQueue.m:151)
6  Foundation                     0x1adddc470 __NSFireDelayedPerform + 408
7  CoreFoundation                 0x1ad2fa8f4 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 28
8  CoreFoundation                 0x1ad2fa624 __CFRunLoopDoTimer + 864
9  CoreFoundation                 0x1ad2f9e58 __CFRunLoopDoTimers + 248
10 CoreFoundation                 0x1ad2f4da8 __CFRunLoopRun + 1844
11 CoreFoundation                 0x1ad2f4354 CFRunLoopRunSpecific + 436
12 GraphicsServices               0x1af4f479c GSEventRunModal + 104
13 UIKitCore                      0x1d979db68 UIApplicationMain + 212
14 TheatreEars                    0x104812830 main (main.m:14)
15 libdyld.dylib                  0x1acdba8e0 start + 4

**More Evidence**:

Crashed: com.apple.main-thread
0  libobjc.A.dylib                0x1952afbe8 objc_retain + 8
1  TheatreEars                    0x10462d824 -[MPURLRequest initWithURL:] (MPURLRequest.m:59)
2  TheatreEars                    0x104739d48 -[MPAnalyticsTracker sendTrackingRequestForURLs:] (MPAnalyticsTracker.m:46)
3  TheatreEars                    0x104739abc -[MPAnalyticsTracker trackImpressionForConfiguration:] (MPAnalyticsTracker.m:33)
4  TheatreEars                    0x10468c1f0 -[MPBaseInterstitialAdapter trackImpression] (MPBaseInterstitialAdapter.m:107)
5  TheatreEars                    0x104648e68 -[MPInterstitialCustomEventAdapter trackImpression] (MPInterstitialCustomEventAdapter.m:168)
6  TheatreEars                    0x104648c00 -[MPInterstitialCustomEventAdapter interstitialCustomEventDidAppear:] (MPInterstitialCustomEventAdapter.m:131)
7  TheatreEars                    0x10456e56c -[MPMRAIDInterstitialCustomEvent interstitialDidAppear:] (MPMRAIDInterstitialCustomEvent.m:84)
8  TheatreEars                    0x1045533e8 __59-[MPMRAIDInterstitialViewController didPresentInterstitial]_block_invoke (MPMRAIDInterstitialViewController.m:68)
9  libdispatch.dylib              0x195afca38 _dispatch_call_block_and_release + 24
10 libdispatch.dylib              0x195afd7d4 _dispatch_client_callout + 16
11 libdispatch.dylib              0x195aab004 _dispatch_main_queue_callback_4CF$VARIANT$mp + 1068
12 CoreFoundation                 0x19604dec0 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 12
13 CoreFoundation                 0x196048df8 __CFRunLoopRun + 1924
14 CoreFoundation                 0x196048354 CFRunLoopRunSpecific + 436
15 GraphicsServices               0x19824879c GSEventRunModal + 104
16 UIKitCore                      0x1c24bbb68 UIApplicationMain + 212
17 TheatreEars                    0x104686830 main (main.m:14)
18 libdyld.dylib                  0x195b0e8e0 start + 4

**Comment by chauduyphanvu1** [comment]
> I haven't used CocaPods to integrate MoPub, instead manually added MoPub folder to my project by downloading from git. Do you think this could also be one of the reason?

It shouldn't be a reason for this crash. That integration option is a supported one.

> When I am using my test devices, the debugger is able to log MoPub UserAgent in the console. I'll try and figure out a way to log this on the live build.

It makes sense because your devices have proper user agent string. The fact that the crashes are only happening in production might be due to a subset of iPhones not returning the user agent properly. Thank you for adding the log to your live build. Let us know once you have the data.

> Also I noticed in my Crashlytics, that there are multiple different session where the crash from somewhere else such as requesting Native Ads or Interstitial Ads starts but ends up at [MPURLRequest initWithUrl].

Hmm, that's not expected that some other code paths also crash here while the SDK processes the user agent. Let me look into the call stack in your trace a bit more and get back if I have any findings.

**Comment by rsatrasala** [comment]
Not Reproducible. Closing.
Please open a new issue if needed.
